reflex-dom-core/src/Reflex/Dom/Attributes/Types.hs

```diff
@@ -0,0 +1,298 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Types for creating or interacting with attributes.
+module Reflex.Dom.Attributes.Types
+  (
+  -- * Base types
+    Namespace
+  , AttributeName(..)
+
+  -- * Text based attribute functions
+  , declareAttribute
+  , setAttribute
+  , removeAttribute
+  , declareAttributeMap
+  , mapKeysToAttributeName
+
+  -- * Attributes
+  , DeclareAttrs
+  , attributesPatchIsEmpty
+  , removeAttrFromDeclareAttrs
+  , lookupAttrInPatch
+  , ModifyAttrs
+  , removeAttrFromModifyAttrs
+  , toPatch
+
+  -- * Implementing new attributes
+  , IsAttribute(..)
+  , PatchSettings(..)
+  , singleAttribute
+  , singleModifyAttribute
+  ) where
+
+import Control.Lens
+import Control.Monad (when, void)
+import Data.Dependent.Map (DMap)
+import Data.Dependent.Sum (DSum(..))
+import Data.Foldable (traverse_)
+import Data.GADT.Compare.TH (deriveGEq, deriveGCompare)
+import Data.Map (Map)
+import Data.Map.Misc (diffMap)
+import Data.Patch
+import Data.Proxy
+import Data.String (IsString(..))
+import Data.Text
+import GHCJS.DOM.Types (JSM)
+import Type.Reflection
+import qualified Data.Dependent.Map as DMap
+import qualified Data.Dependent.Map.Lens as DMap
+import qualified Data.Map as Map
+import qualified GHCJS.DOM.Element as Element
+import qualified GHCJS.DOM.Types as DOM
+
+type Namespace = Text
+
+-- | Attribute names which may or may not have namespaces
+data AttributeName = AttributeName !(Maybe Namespace) !Text deriving (Show, Read, Eq, Ord)
+
+-- | By default, AttributeNames are unnamespaced
+instance IsString AttributeName where
+  fromString = AttributeName Nothing . fromString
+
+-- | Used for controlling 'applyAttrPatchDOM'.
+data PatchSettings
+  = PatchSettings_CheckBeforePatching
+  -- ^ Be careful to minimise the patch. This is used for checking attrs during
+  -- hydration, to prevent touching the DOM when unnecessary. In particular,
+  -- always patching during hydration can result in CSS animations running more
+  -- than once.
+  | PatchSettings_PatchAlways
+  -- ^ Always run the patch. Alternatively: checking before patching is
+  -- optional, and can be omitted to avoid an extra roundtrip.
+
+-- | Define new attributes by creating instances of this class.
+-- This is defined around 'Patch' so we can use 'holdIncremental'.
+-- The 'PatchTarget' is used in declarative attributes (i.e. static and
+-- 'Dynamic' based attributes), and the patch itself is used in modify
+-- attributes ('Event' based attributes).
+class (Monoid (PatchTarget a), Semigroup a, Patch a) => IsAttribute a where
```

Comment:
The methods of this class only make sense in the context of knowing about the Hydration, Immediate, and Static dom builders, as well as the way that `elDynAttr` works.  The `DomBuilder` class depends on `IsAttribute`, and `IsAttribute` logically depends on knowing all the *instances* of `DomBuilder`.  In essence, this makes the DomBuilder class not useful as a class: we could instead just merge all these implementations and let the user choose with a sum type.